[BUG] hash_aggregate_test TypedImperativeAggregate tests failed

**Describe the bug**
tests added in https://github.com/NVIDIA/spark-rapids/pull/2971. We first saw this in databricks runtime 7.3 and 8.2 (nightly pipeline)

**Caused by: ai.rapids.cudf.CudfException: vector::_M_range_check: __n (which is 4) >= this->size() (which is 4)**

failures list,
```
[2021-08-04T05:18:30.416Z] FAILED ../../src/main/python/hash_aggregate_test.py::test_hash_groupby_collect_with_single_distinct[[('a', RepeatSeq(Long)), ('b', RepeatSeq(Boolean)), ('c', LongRange(not_null))]][IGNORE_ORDER({'local': True}), INCOMPAT, APPROXIMATE_FLOAT]
[2021-08-04T05:18:30.416Z] FAILED ../../src/main/python/hash_aggregate_test.py::test_hash_groupby_collect_with_single_distinct[[('a', RepeatSeq(Long)), ('b', RepeatSeq(Integer)), ('c', LongRange(not_null))]][IGNORE_ORDER({'local': True}), INCOMPAT, APPROXIMATE_FLOAT]
[2021-08-04T05:18:30.416Z] FAILED ../../src/main/python/hash_aggregate_test.py::test_hash_groupby_collect_with_single_distinct[[('a', RepeatSeq(Long)), ('b', RepeatSeq(Long)), ('c', LongRange(not_null))]][IGNORE_ORDER({'local': True}), INCOMPAT, APPROXIMATE_FLOAT]
[2021-08-04T05:18:30.416Z] FAILED ../../src/main/python/hash_aggregate_test.py::test_hash_groupby_collect_with_single_distinct[[('a', RepeatSeq(Long)), ('b', RepeatSeq(Short)), ('c', LongRange(not_null))]][IGNORE_ORDER({'local': True}), INCOMPAT, APPROXIMATE_FLOAT]
[2021-08-04T05:18:30.416Z] FAILED ../../src/main/python/hash_aggregate_test.py::test_hash_groupby_collect_with_single_distinct[[('a', RepeatSeq(Long)), ('b', RepeatSeq(Date)), ('c', LongRange(not_null))]][IGNORE_ORDER({'local': True}), INCOMPAT, APPROXIMATE_FLOAT]
[2021-08-04T05:18:30.416Z] FAILED ../../src/main/python/hash_aggregate_test.py::test_hash_groupby_collect_with_single_distinct[[('a', RepeatSeq(Long)), ('b', RepeatSeq(Timestamp)), ('c', LongRange(not_null))]][IGNORE_ORDER({'local': True}), INCOMPAT, APPROXIMATE_FLOAT]
[2021-08-04T05:18:30.417Z] FAILED ../../src/main/python/hash_aggregate_test.py::test_hash_groupby_collect_with_single_distinct[[('a', RepeatSeq(Long)), ('b', RepeatSeq(Byte)), ('c', LongRange(not_null))]][IGNORE_ORDER({'local': True}), INCOMPAT, APPROXIMATE_FLOAT]
[2021-08-04T05:18:30.417Z] FAILED ../../src/main/python/hash_aggregate_test.py::test_hash_groupby_collect_with_single_distinct[[('a', RepeatSeq(Long)), ('b', RepeatSeq(String)), ('c', LongRange(not_null))]][IGNORE_ORDER({'local': True}), INCOMPAT, APPROXIMATE_FLOAT]
[2021-08-04T05:18:30.417Z] FAILED ../../src/main/python/hash_aggregate_test.py::test_hash_groupby_collect_with_single_distinct[[('a', RepeatSeq(Long)), ('b', RepeatSeq(Float)), ('c', LongRange(not_null))]0][IGNORE_ORDER({'local': True}), INCOMPAT, APPROXIMATE_FLOAT]
[2021-08-04T05:18:30.417Z] FAILED ../../src/main/python/hash_aggregate_test.py::test_hash_groupby_collect_with_single_distinct[[('a', RepeatSeq(Long)), ('b', RepeatSeq(Double)), ('c', LongRange(not_null))]][IGNORE_ORDER({'local': True}), INCOMPAT, APPROXIMATE_FLOAT]
[2021-08-04T05:18:30.417Z] FAILED ../../src/main/python/hash_aggregate_test.py::test_hash_groupby_collect_with_single_distinct[[('a', RepeatSeq(Long)), ('b', RepeatSeq(Decimal(8,3))), ('c', LongRange(not_null))]][IGNORE_ORDER({'local': True}), INCOMPAT, APPROXIMATE_FLOAT]
[2021-08-04T05:18:30.417Z] FAILED ../../src/main/python/hash_aggregate_test.py::test_hash_groupby_collect_with_single_distinct[[('a', RepeatSeq(Long)), ('b', RepeatSeq(Float)), ('c', LongRange(not_null))]1][IGNORE_ORDER({'local': True}), INCOMPAT, APPROXIMATE_FLOAT]
[2021-08-04T05:18:30.417Z] FAILED ../../src/main/python/hash_aggregate_test.py::test_hash_groupby_single_distinct_collect[[('a', RepeatSeq(Long)), ('b', RepeatSeq(Boolean)), ('c', LongRange(not_null))]][IGNORE_ORDER({'local': True}), INCOMPAT, APPROXIMATE_FLOAT]
[2021-08-04T05:18:30.417Z] FAILED ../../src/main/python/hash_aggregate_test.py::test_hash_groupby_single_distinct_collect[[('a', RepeatSeq(Long)), ('b', RepeatSeq(Integer)), ('c', LongRange(not_null))]][IGNORE_ORDER({'local': True}), INCOMPAT, APPROXIMATE_FLOAT]
[2021-08-04T05:18:30.417Z] FAILED ../../src/main/python/hash_aggregate_test.py::test_hash_groupby_single_distinct_collect[[('a', RepeatSeq(Long)), ('b', RepeatSeq(Long)), ('c', LongRange(not_null))]][IGNORE_ORDER({'local': True}), INCOMPAT, APPROXIMATE_FLOAT]
[2021-08-04T05:18:30.417Z] FAILED ../../src/main/python/hash_aggregate_test.py::test_hash_groupby_single_distinct_collect[[('a', RepeatSeq(Long)), ('b', RepeatSeq(Short)), ('c', LongRange(not_null))]][IGNORE_ORDER({'local': True}), INCOMPAT, APPROXIMATE_FLOAT]
[2021-08-04T05:18:30.417Z] FAILED ../../src/main/python/hash_aggregate_test.py::test_hash_groupby_single_distinct_collect[[('a', RepeatSeq(Long)), ('b', RepeatSeq(Date)), ('c', LongRange(not_null))]][IGNORE_ORDER({'local': True}), INCOMPAT, APPROXIMATE_FLOAT]
[2021-08-04T05:18:30.417Z] FAILED ../../src/main/python/hash_aggregate_test.py::test_hash_groupby_single_distinct_collect[[('a', RepeatSeq(Long)), ('b', RepeatSeq(Timestamp)), ('c', LongRange(not_null))]][IGNORE_ORDER({'local': True}), INCOMPAT, APPROXIMATE_FLOAT]
[2021-08-04T05:18:30.417Z] FAILED ../../src/main/python/hash_aggregate_test.py::test_hash_groupby_single_distinct_collect[[('a', RepeatSeq(Long)), ('b', RepeatSeq(Byte)), ('c', LongRange(not_null))]][IGNORE_ORDER({'local': True}), INCOMPAT, APPROXIMATE_FLOAT]
[2021-08-04T05:18:30.417Z] FAILED ../../src/main/python/hash_aggregate_test.py::test_hash_groupby_single_distinct_collect[[('a', RepeatSeq(Long)), ('b', RepeatSeq(String)), ('c', LongRange(not_null))]][IGNORE_ORDER({'local': True}), INCOMPAT, APPROXIMATE_FLOAT]
[2021-08-04T05:18:30.417Z] FAILED ../../src/main/python/hash_aggregate_test.py::test_hash_groupby_single_distinct_collect[[('a', RepeatSeq(Long)), ('b', RepeatSeq(Float)), ('c', LongRange(not_null))]0][IGNORE_ORDER({'local': True}), INCOMPAT, APPROXIMATE_FLOAT]
[2021-08-04T05:18:30.417Z] FAILED ../../src/main/python/hash_aggregate_test.py::test_hash_groupby_single_distinct_collect[[('a', RepeatSeq(Long)), ('b', RepeatSeq(Double)), ('c', LongRange(not_null))]][IGNORE_ORDER({'local': True}), INCOMPAT, APPROXIMATE_FLOAT]
[2021-08-04T05:18:30.417Z] FAILED ../../src/main/python/hash_aggregate_test.py::test_hash_groupby_single_distinct_collect[[('a', RepeatSeq(Long)), ('b', RepeatSeq(Decimal(8,3))), ('c', LongRange(not_null))]][IGNORE_ORDER({'local': True}), INCOMPAT, APPROXIMATE_FLOAT]
[2021-08-04T05:18:30.417Z] FAILED ../../src/main/python/hash_aggregate_test.py::test_hash_groupby_single_distinct_collect[[('a', RepeatSeq(Long)), ('b', RepeatSeq(Float)), ('c', LongRange(not_null))]1][IGNORE_ORDER({'local': True}), INCOMPAT, APPROXIMATE_FLOAT]
```

detailed log,
```
[2021-08-04T05:18:30.156Z] ----------------------------- Captured stdout call -----------------------------
[2021-08-04T05:18:30.156Z] ### CPU RUN ###
[2021-08-04T05:18:30.156Z] ### GPU RUN ###
[2021-08-04T05:18:30.156Z] [31m[1m_ test_hash_groupby_single_distinct_collect[[('a', RepeatSeq(Long)), ('b', RepeatSeq(Float)), ('c', LongRange(not_null))]1] _[0m
[2021-08-04T05:18:30.156Z] [gw4] linux -- Python 3.7.10 /databricks/conda/envs/databricks-ml-gpu/bin/python
[2021-08-04T05:18:30.156Z] 
[2021-08-04T05:18:30.156Z] data_gen = [('a', RepeatSeq(Long)), ('b', RepeatSeq(Float)), ('c', LongRange(not_null))]
[2021-08-04T05:18:30.156Z] 
[2021-08-04T05:18:30.156Z]     [37m@approximate_float[39;49;00m
[2021-08-04T05:18:30.156Z]     [37m@ignore_order[39;49;00m(local=[94mTrue[39;49;00m)
[2021-08-04T05:18:30.156Z]     [37m@incompat[39;49;00m
[2021-08-04T05:18:30.156Z]     [37m@pytest[39;49;00m.mark.parametrize([33m'[39;49;00m[33mdata_gen[39;49;00m[33m'[39;49;00m, _gen_data_for_collect_op, ids=idfn)
[2021-08-04T05:18:30.156Z]     [94mdef[39;49;00m [92mtest_hash_groupby_single_distinct_collect[39;49;00m(data_gen):
[2021-08-04T05:18:30.156Z]         [90m# test distinct collect with other aggregations[39;49;00m
[2021-08-04T05:18:30.156Z]         sql = [33m"""[39;49;00m[33mselect a,[39;49;00m[33m[39;49;00m
[2021-08-04T05:18:30.156Z]     [33m                    sort_array(collect_list(distinct b)),[39;49;00m[33m[39;49;00m
[2021-08-04T05:18:30.156Z]     [33m                    sort_array(collect_set(b)),[39;49;00m[33m[39;49;00m
[2021-08-04T05:18:30.156Z]     [33m                    count(distinct b),[39;49;00m[33m[39;49;00m
[2021-08-04T05:18:30.156Z]     [33m                    count(c)[39;49;00m[33m[39;49;00m
[2021-08-04T05:18:30.156Z]     [33m            from tbl group by a[39;49;00m[33m"""[39;49;00m
[2021-08-04T05:18:30.156Z]         assert_gpu_and_cpu_are_equal_sql(
[2021-08-04T05:18:30.156Z]             df_fun=[94mlambda[39;49;00m spark: gen_df(spark, data_gen, length=[94m100[39;49;00m),
[2021-08-04T05:18:30.156Z] >           table_name=[33m"[39;49;00m[33mtbl[39;49;00m[33m"[39;49;00m, sql=sql)
[2021-08-04T05:18:30.156Z] 
[2021-08-04T05:18:30.156Z] [1m[31m../../src/main/python/hash_aggregate_test.py[0m:455: 
[2021-08-04T05:18:30.156Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[2021-08-04T05:18:30.156Z] [1m[31m../../src/main/python/asserts.py[0m:476: in assert_gpu_and_cpu_are_equal_sql
[2021-08-04T05:18:30.156Z]     assert_gpu_and_cpu_are_equal_collect(do_it_all, conf, is_cpu_first=is_cpu_first)
[2021-08-04T05:18:30.156Z] [1m[31m../../src/main/python/asserts.py[0m:434: in assert_gpu_and_cpu_are_equal_collect
[2021-08-04T05:18:30.156Z]     _assert_gpu_and_cpu_are_equal(func, [33m'[39;49;00m[33mCOLLECT[39;49;00m[33m'[39;49;00m, conf=conf, is_cpu_first=is_cpu_first)
[2021-08-04T05:18:30.156Z] [1m[31m../../src/main/python/asserts.py[0m:415: in _assert_gpu_and_cpu_are_equal
[2021-08-04T05:18:30.156Z]     run_on_gpu()
[2021-08-04T05:18:30.156Z] [1m[31m../../src/main/python/asserts.py[0m:409: in run_on_gpu
[2021-08-04T05:18:30.156Z]     from_gpu = with_gpu_session(bring_back, conf=conf)
[2021-08-04T05:18:30.156Z] [1m[31m../../src/main/python/spark_session.py[0m:105: in with_gpu_session
[2021-08-04T05:18:30.156Z]     [94mreturn[39;49;00m with_spark_session(func, conf=copy)
[2021-08-04T05:18:30.157Z] [1m[31m../../src/main/python/spark_session.py[0m:70: in with_spark_session
[2021-08-04T05:18:30.157Z]     ret = func(_spark)
[2021-08-04T05:18:30.157Z] [1m[31m../../src/main/python/asserts.py[0m:190: in <lambda>
[2021-08-04T05:18:30.157Z]     bring_back = [94mlambda[39;49;00m spark: limit_func(spark).collect()
[2021-08-04T05:18:30.157Z] [1m[31m/databricks/spark/python/pyspark/sql/dataframe.py[0m:611: in collect
[2021-08-04T05:18:30.157Z]     sock_info = [96mself[39;49;00m._jdf.collectToPython()
[2021-08-04T05:18:30.157Z] [1m[31m/databricks/spark/python/lib/py4j-0.10.9-src.zip/py4j/java_gateway.py[0m:1305: in __call__
[2021-08-04T05:18:30.157Z]     answer, [96mself[39;49;00m.gateway_client, [96mself[39;49;00m.target_id, [96mself[39;49;00m.name)
[2021-08-04T05:18:30.157Z] [1m[31m/databricks/spark/python/pyspark/sql/utils.py[0m:127: in deco
[2021-08-04T05:18:30.157Z]     [94mreturn[39;49;00m f(*a, **kw)
[2021-08-04T05:18:30.157Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[2021-08-04T05:18:30.157Z] 
[2021-08-04T05:18:30.157Z] answer = 'xro192091'
[2021-08-04T05:18:30.157Z] gateway_client = <py4j.java_gateway.GatewayClient object at 0x7f5cc49f7450>
[2021-08-04T05:18:30.157Z] target_id = 'o192088', name = 'collectToPython'
[2021-08-04T05:18:30.157Z] 
[2021-08-04T05:18:30.413Z]     [94mdef[39;49;00m [92mget_return_value[39;49;00m(answer, gateway_client, target_id=[94mNone[39;49;00m, name=[94mNone[39;49;00m):
[2021-08-04T05:18:30.413Z]         [33m"""Converts an answer received from the Java gateway into a Python object.[39;49;00m
[2021-08-04T05:18:30.413Z]     [33m[39;49;00m
[2021-08-04T05:18:30.413Z]     [33m    For example, string representation of integers are converted to Python[39;49;00m
[2021-08-04T05:18:30.413Z]     [33m    integer, string representation of objects are converted to JavaObject[39;49;00m
[2021-08-04T05:18:30.413Z]     [33m    instances, etc.[39;49;00m
[2021-08-04T05:18:30.413Z]     [33m[39;49;00m
[2021-08-04T05:18:30.413Z]     [33m    :param answer: the string returned by the Java gateway[39;49;00m
[2021-08-04T05:18:30.413Z]     [33m    :param gateway_client: the gateway client used to communicate with the Java[39;49;00m
[2021-08-04T05:18:30.413Z]     [33m        Gateway. Only necessary if the answer is a reference (e.g., object,[39;49;00m
[2021-08-04T05:18:30.413Z]     [33m        list, map)[39;49;00m
[2021-08-04T05:18:30.413Z]     [33m    :param target_id: the name of the object from which the answer comes from[39;49;00m
[2021-08-04T05:18:30.413Z]     [33m        (e.g., *object1* in `object1.hello()`). Optional.[39;49;00m
[2021-08-04T05:18:30.413Z]     [33m    :param name: the name of the member from which the answer comes from[39;49;00m
[2021-08-04T05:18:30.413Z]     [33m        (e.g., *hello* in `object1.hello()`). Optional.[39;49;00m
[2021-08-04T05:18:30.413Z]     [33m    """[39;49;00m
[2021-08-04T05:18:30.413Z]         [94mif[39;49;00m is_error(answer)[[94m0[39;49;00m]:
[2021-08-04T05:18:30.413Z]             [94mif[39;49;00m [96mlen[39;49;00m(answer) > [94m1[39;49;00m:
[2021-08-04T05:18:30.413Z]                 [96mtype[39;49;00m = answer[[94m1[39;49;00m]
[2021-08-04T05:18:30.413Z]                 value = OUTPUT_CONVERTER[[96mtype[39;49;00m](answer[[94m2[39;49;00m:], gateway_client)
[2021-08-04T05:18:30.413Z]                 [94mif[39;49;00m answer[[94m1[39;49;00m] == REFERENCE_TYPE:
[2021-08-04T05:18:30.413Z]                     [94mraise[39;49;00m Py4JJavaError(
[2021-08-04T05:18:30.413Z]                         [33m"[39;49;00m[33mAn error occurred while calling [39;49;00m[33m{0}[39;49;00m[33m{1}[39;49;00m[33m{2}[39;49;00m[33m.[39;49;00m[33m\n[39;49;00m[33m"[39;49;00m.
[2021-08-04T05:18:30.413Z] >                       [96mformat[39;49;00m(target_id, [33m"[39;49;00m[33m.[39;49;00m[33m"[39;49;00m, name), value)
[2021-08-04T05:18:30.413Z] [1m[31mE                   py4j.protocol.Py4JJavaError: An error occurred while calling o192088.collectToPython.[0m
[2021-08-04T05:18:30.413Z] [1m[31mE                   : org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 6982.0 failed 4 times, most recent failure: Lost task 0.3 in stage 6982.0 (TID 33637, ip-10-59-181-194.us-west-2.compute.internal, executor driver): ai.rapids.cudf.CudfException: vector::_M_range_check: __n (which is 4) >= this->size() (which is 4)[0m
[2021-08-04T05:18:30.413Z] [1m[31mE                   	at ai.rapids.cudf.Table.groupByAggregate(Native Method)[0m
[2021-08-04T05:18:30.413Z] [1m[31mE                   	at ai.rapids.cudf.Table.access$3000(Table.java:39)[0m
[2021-08-04T05:18:30.413Z] [1m[31mE                   	at ai.rapids.cudf.Table$GroupByOperation.aggregate(Table.java:2490)[0m
[2021-08-04T05:18:30.413Z] [1m[31mE                   	at com.nvidia.spark.rapids.GpuHashAggregateIterator.$anonfun$computeAggregate$7(aggregate.scala:762)[0m
[2021-08-04T05:18:30.413Z] [1m[31mE                   	at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)[0m
[2021-08-04T05:18:30.413Z] [1m[31mE                   	at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)[0m
[2021-08-04T05:18:30.413Z] [1m[31mE                   	at com.nvidia.spark.rapids.GpuHashAggregateIterator.withResource(aggregate.scala:209)[0m
[2021-08-04T05:18:30.413Z] [1m[31mE                   	at com.nvidia.spark.rapids.GpuHashAggregateIterator.$anonfun$computeAggregate$1(aggregate.scala:761)[0m
[2021-08-04T05:18:30.413Z] [1m[31mE                   	at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)[0m
[2021-08-04T05:18:30.413Z] [1m[31mE                   	at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)[0m
[2021-08-04T05:18:30.413Z] [1m[31mE                   	at com.nvidia.spark.rapids.GpuHashAggregateIterator.withResource(aggregate.scala:209)[0m
[2021-08-04T05:18:30.413Z] [1m[31mE                   	at com.nvidia.spark.rapids.GpuHashAggregateIterator.com$nvidia$spark$rapids$GpuHashAggregateIterator$$computeAggregate(aggregate.scala:740)[0m
[2021-08-04T05:18:30.413Z] [1m[31mE                   	at com.nvidia.spark.rapids.GpuHashAggregateIterator.$anonfun$aggregateInputBatches$2(aggregate.scala:319)[0m
[2021-08-04T05:18:30.413Z] [1m[31mE                   	at com.nvidia.spark.rapids.GpuHashAggregateIterator.$anonfun$aggregateInputBatches$2$adapted(aggregate.scala:318)[0m
[2021-08-04T05:18:30.413Z] [1m[31mE                   	at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:46)[0m
[2021-08-04T05:18:30.413Z] [1m[31mE                   	at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:44)[0m
[2021-08-04T05:18:30.413Z] [1m[31mE                   	at com.nvidia.spark.rapids.GpuHashAggregateIterator.withResource(aggregate.scala:209)[0m
[2021-08-04T05:18:30.413Z] [1m[31mE                   	at com.nvidia.spark.rapids.GpuHashAggregateIterator.aggregateInputBatches(aggregate.scala:318)[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   	at com.nvidia.spark.rapids.GpuHashAggregateIterator.$anonfun$next$2(aggregate.scala:269)[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   	at scala.Option.getOrElse(Option.scala:189)[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   	at com.nvidia.spark.rapids.GpuHashAggregateIterator.next(aggregate.scala:266)[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   	at com.nvidia.spark.rapids.GpuHashAggregateIterator.next(aggregate.scala:209)[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   	at com.nvidia.spark.rapids.ColumnarToRowIterator.$anonfun$fetchNextBatch$2(GpuColumnarToRowExec.scala:224)[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   	at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   	at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   	at com.nvidia.spark.rapids.ColumnarToRowIterator.withResource(GpuColumnarToRowExec.scala:177)[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   	at com.nvidia.spark.rapids.ColumnarToRowIterator.fetchNextBatch(GpuColumnarToRowExec.scala:221)[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   	at com.nvidia.spark.rapids.ColumnarToRowIterator.loadNextBatch(GpuColumnarToRowExec.scala:198)[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   	at com.nvidia.spark.rapids.ColumnarToRowIterator.hasNext(GpuColumnarToRowExec.scala:238)[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:458)[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   	at org.apache.spark.sql.execution.collect.UnsafeRowBatchUtils$.encodeUnsafeRows(UnsafeRowBatchUtils.scala:80)[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   	at org.apache.spark.sql.execution.collect.Collector.$anonfun$processPartition$1(Collector.scala:179)[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   	at org.apache.spark.SparkContext.$anonfun$runJob$6(SparkContext.scala:2433)[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   	at org.apache.spark.scheduler.Task.doRunTask(Task.scala:144)[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   	at org.apache.spark.scheduler.Task.run(Task.scala:117)[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$9(Executor.scala:640)[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1581)[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:643)[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   	at java.lang.Thread.run(Thread.java:748)[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   [0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   Driver stacktrace:[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   	at org.apache.spark.scheduler.DAGScheduler.failJobAndIndependentStages(DAGScheduler.scala:2519)[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   	at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2(DAGScheduler.scala:2466)[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   	at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2$adapted(DAGScheduler.scala:2460)[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   	at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:2460)[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   	at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1(DAGScheduler.scala:1152)[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   	at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1$adapted(DAGScheduler.scala:1152)[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   	at scala.Option.foreach(Option.scala:407)[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   	at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:1152)[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:2721)[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2668)[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2656)[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   	at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   	at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:938)[0m
[2021-08-04T05:18:30.414Z] [1m[31mE                   	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2339)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2434)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at org.apache.spark.sql.execution.collect.Collector.runSparkJobs(Collector.scala:273)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at org.apache.spark.sql.execution.collect.Collector.collect(Collector.scala:308)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at org.apache.spark.sql.execution.collect.Collector$.collect(Collector.scala:82)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at org.apache.spark.sql.execution.collect.Collector$.collect(Collector.scala:88)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at org.apache.spark.sql.execution.ResultCacheManager.getOrComputeResult(ResultCacheManager.scala:508)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at org.apache.spark.sql.execution.ResultCacheManager.getOrComputeResult(ResultCacheManager.scala:480)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at org.apache.spark.sql.execution.SparkPlan.executeCollectResult(SparkPlan.scala:401)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at org.apache.spark.sql.Dataset.$anonfun$collectToPython$1(Dataset.scala:3497)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at org.apache.spark.sql.Dataset.$anonfun$withAction$1(Dataset.scala:3709)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withCustomExecutionEnv$5(SQLExecution.scala:116)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:249)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withCustomExecutionEnv$1(SQLExecution.scala:101)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:845)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at org.apache.spark.sql.execution.SQLExecution$.withCustomExecutionEnv(SQLExecution.scala:77)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:199)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at org.apache.spark.sql.Dataset.withAction(Dataset.scala:3707)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at org.apache.spark.sql.Dataset.collectToPython(Dataset.scala:3495)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at sun.reflect.GeneratedMethodAccessor130.invoke(Unknown Source)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at java.lang.reflect.Method.invoke(Method.java:498)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:380)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at py4j.Gateway.invoke(Gateway.java:295)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at py4j.commands.CallCommand.execute(CallCommand.java:79)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at py4j.GatewayConnection.run(GatewayConnection.java:251)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at java.lang.Thread.run(Thread.java:748)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   Caused by: ai.rapids.cudf.CudfException: vector::_M_range_check: __n (which is 4) >= this->size() (which is 4)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at ai.rapids.cudf.Table.groupByAggregate(Native Method)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at ai.rapids.cudf.Table.access$3000(Table.java:39)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at ai.rapids.cudf.Table$GroupByOperation.aggregate(Table.java:2490)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at com.nvidia.spark.rapids.GpuHashAggregateIterator.$anonfun$computeAggregate$7(aggregate.scala:762)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at com.nvidia.spark.rapids.GpuHashAggregateIterator.withResource(aggregate.scala:209)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at com.nvidia.spark.rapids.GpuHashAggregateIterator.$anonfun$computeAggregate$1(aggregate.scala:761)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at com.nvidia.spark.rapids.GpuHashAggregateIterator.withResource(aggregate.scala:209)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at com.nvidia.spark.rapids.GpuHashAggregateIterator.com$nvidia$spark$rapids$GpuHashAggregateIterator$$computeAggregate(aggregate.scala:740)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at com.nvidia.spark.rapids.GpuHashAggregateIterator.$anonfun$aggregateInputBatches$2(aggregate.scala:319)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at com.nvidia.spark.rapids.GpuHashAggregateIterator.$anonfun$aggregateInputBatches$2$adapted(aggregate.scala:318)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:46)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:44)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at com.nvidia.spark.rapids.GpuHashAggregateIterator.withResource(aggregate.scala:209)[0m
[2021-08-04T05:18:30.415Z] [1m[31mE                   	at com.nvidia.spark.rapids.GpuHashAggregateIterator.aggregateInputBatches(aggregate.scala:318)[0m
[2021-08-04T05:18:30.416Z] [1m[31mE                   	at com.nvidia.spark.rapids.GpuHashAggregateIterator.$anonfun$next$2(aggregate.scala:269)[0m
[2021-08-04T05:18:30.416Z] [1m[31mE                   	at scala.Option.getOrElse(Option.scala:189)[0m
[2021-08-04T05:18:30.416Z] [1m[31mE                   	at com.nvidia.spark.rapids.GpuHashAggregateIterator.next(aggregate.scala:266)[0m
[2021-08-04T05:18:30.416Z] [1m[31mE                   	at com.nvidia.spark.rapids.GpuHashAggregateIterator.next(aggregate.scala:209)[0m
[2021-08-04T05:18:30.416Z] [1m[31mE                   	at com.nvidia.spark.rapids.ColumnarToRowIterator.$anonfun$fetchNextBatch$2(GpuColumnarToRowExec.scala:224)[0m
[2021-08-04T05:18:30.416Z] [1m[31mE                   	at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)[0m
[2021-08-04T05:18:30.416Z] [1m[31mE                   	at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)[0m
[2021-08-04T05:18:30.416Z] [1m[31mE                   	at com.nvidia.spark.rapids.ColumnarToRowIterator.withResource(GpuColumnarToRowExec.scala:177)[0m
[2021-08-04T05:18:30.416Z] [1m[31mE                   	at com.nvidia.spark.rapids.ColumnarToRowIterator.fetchNextBatch(GpuColumnarToRowExec.scala:221)[0m
[2021-08-04T05:18:30.416Z] [1m[31mE                   	at com.nvidia.spark.rapids.ColumnarToRowIterator.loadNextBatch(GpuColumnarToRowExec.scala:198)[0m
[2021-08-04T05:18:30.416Z] [1m[31mE                   	at com.nvidia.spark.rapids.ColumnarToRowIterator.hasNext(GpuColumnarToRowExec.scala:238)[0m
[2021-08-04T05:18:30.416Z] [1m[31mE                   	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:458)[0m
[2021-08-04T05:18:30.416Z] [1m[31mE                   	at org.apache.spark.sql.execution.collect.UnsafeRowBatchUtils$.encodeUnsafeRows(UnsafeRowBatchUtils.scala:80)[0m
[2021-08-04T05:18:30.416Z] [1m[31mE                   	at org.apache.spark.sql.execution.collect.Collector.$anonfun$processPartition$1(Collector.scala:179)[0m
[2021-08-04T05:18:30.416Z] [1m[31mE                   	at org.apache.spark.SparkContext.$anonfun$runJob$6(SparkContext.scala:2433)[0m
[2021-08-04T05:18:30.416Z] [1m[31mE                   	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)[0m
[2021-08-04T05:18:30.416Z] [1m[31mE                   	at org.apache.spark.scheduler.Task.doRunTask(Task.scala:144)[0m
[2021-08-04T05:18:30.416Z] [1m[31mE                   	at org.apache.spark.scheduler.Task.run(Task.scala:117)[0m
[2021-08-04T05:18:30.416Z] [1m[31mE                   	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$9(Executor.scala:640)[0m
[2021-08-04T05:18:30.416Z] [1m[31mE                   	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1581)[0m
[2021-08-04T05:18:30.416Z] [1m[31mE                   	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:643)[0m
[2021-08-04T05:18:30.416Z] [1m[31mE                   	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)[0m
[2021-08-04T05:18:30.416Z] [1m[31mE                   	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)[0m
[2021-08-04T05:18:30.416Z] [1m[31mE                   	... 1 more[0m
```

**UPDATE:**
we have 12 more failures for databricks 8.2 runtime and spark 3.0.3+

**pyspark.sql.utils.IllegalArgumentException: The expression coalesce(first(if ((gid#1738853 = 0)) collect_list(`b`)#1738857 else null) ignore nulls#1738858, []) is not columnar class org.apache.spark.sql.catalyst.expressions.Coalesce**

failures list,
```
[2021-08-04T06:38:01.432Z] FAILED ../../src/main/python/hash_aggregate_test.py::test_hash_groupby_collect_with_multi_distinct_fallback[[('a', RepeatSeq(Long)), ('b', RepeatSeq(Boolean)), ('c', LongRange(not_null))]][IGNORE_ORDER({'local': True}), INCOMPAT, APPROXIMATE_FLOAT, ALLOW_NON_GPU(SortAggregateExec,SortArray,Alias,Literal,First,If,EqualTo,Count,CollectList,CollectSet,AggregateExpression)]
[2021-08-04T06:38:01.432Z] FAILED ../../src/main/python/hash_aggregate_test.py::test_hash_groupby_collect_with_multi_distinct_fallback[[('a', RepeatSeq(Long)), ('b', RepeatSeq(Integer)), ('c', LongRange(not_null))]][IGNORE_ORDER({'local': True}), INCOMPAT, APPROXIMATE_FLOAT, ALLOW_NON_GPU(SortAggregateExec,SortArray,Alias,Literal,First,If,EqualTo,Count,CollectList,CollectSet,AggregateExpression)]
[2021-08-04T06:38:01.432Z] FAILED ../../src/main/python/hash_aggregate_test.py::test_hash_groupby_collect_with_multi_distinct_fallback[[('a', RepeatSeq(Long)), ('b', RepeatSeq(Long)), ('c', LongRange(not_null))]][IGNORE_ORDER({'local': True}), INCOMPAT, APPROXIMATE_FLOAT, ALLOW_NON_GPU(SortAggregateExec,SortArray,Alias,Literal,First,If,EqualTo,Count,CollectList,CollectSet,AggregateExpression)]
[2021-08-04T06:38:01.432Z] FAILED ../../src/main/python/hash_aggregate_test.py::test_hash_groupby_collect_with_multi_distinct_fallback[[('a', RepeatSeq(Long)), ('b', RepeatSeq(Short)), ('c', LongRange(not_null))]][IGNORE_ORDER({'local': True}), INCOMPAT, APPROXIMATE_FLOAT, ALLOW_NON_GPU(SortAggregateExec,SortArray,Alias,Literal,First,If,EqualTo,Count,CollectList,CollectSet,AggregateExpression)]
[2021-08-04T06:38:01.433Z] FAILED ../../src/main/python/hash_aggregate_test.py::test_hash_groupby_collect_with_multi_distinct_fallback[[('a', RepeatSeq(Long)), ('b', RepeatSeq(Date)), ('c', LongRange(not_null))]][IGNORE_ORDER({'local': True}), INCOMPAT, APPROXIMATE_FLOAT, ALLOW_NON_GPU(SortAggregateExec,SortArray,Alias,Literal,First,If,EqualTo,Count,CollectList,CollectSet,AggregateExpression)]
[2021-08-04T06:38:01.433Z] FAILED ../../src/main/python/hash_aggregate_test.py::test_hash_groupby_collect_with_multi_distinct_fallback[[('a', RepeatSeq(Long)), ('b', RepeatSeq(Timestamp)), ('c', LongRange(not_null))]][IGNORE_ORDER({'local': True}), INCOMPAT, APPROXIMATE_FLOAT, ALLOW_NON_GPU(SortAggregateExec,SortArray,Alias,Literal,First,If,EqualTo,Count,CollectList,CollectSet,AggregateExpression)]
[2021-08-04T06:38:01.433Z] FAILED ../../src/main/python/hash_aggregate_test.py::test_hash_groupby_collect_with_multi_distinct_fallback[[('a', RepeatSeq(Long)), ('b', RepeatSeq(Byte)), ('c', LongRange(not_null))]][IGNORE_ORDER({'local': True}), INCOMPAT, APPROXIMATE_FLOAT, ALLOW_NON_GPU(SortAggregateExec,SortArray,Alias,Literal,First,If,EqualTo,Count,CollectList,CollectSet,AggregateExpression)]
[2021-08-04T06:38:01.433Z] FAILED ../../src/main/python/hash_aggregate_test.py::test_hash_groupby_collect_with_multi_distinct_fallback[[('a', RepeatSeq(Long)), ('b', RepeatSeq(String)), ('c', LongRange(not_null))]][IGNORE_ORDER({'local': True}), INCOMPAT, APPROXIMATE_FLOAT, ALLOW_NON_GPU(SortAggregateExec,SortArray,Alias,Literal,First,If,EqualTo,Count,CollectList,CollectSet,AggregateExpression)]
[2021-08-04T06:38:01.433Z] FAILED ../../src/main/python/hash_aggregate_test.py::test_hash_groupby_collect_with_multi_distinct_fallback[[('a', RepeatSeq(Long)), ('b', RepeatSeq(Float)), ('c', LongRange(not_null))]0][IGNORE_ORDER({'local': True}), INCOMPAT, APPROXIMATE_FLOAT, ALLOW_NON_GPU(SortAggregateExec,SortArray,Alias,Literal,First,If,EqualTo,Count,CollectList,CollectSet,AggregateExpression)]
[2021-08-04T06:38:01.433Z] FAILED ../../src/main/python/hash_aggregate_test.py::test_hash_groupby_collect_with_multi_distinct_fallback[[('a', RepeatSeq(Long)), ('b', RepeatSeq(Double)), ('c', LongRange(not_null))]][IGNORE_ORDER({'local': True}), INCOMPAT, APPROXIMATE_FLOAT, ALLOW_NON_GPU(SortAggregateExec,SortArray,Alias,Literal,First,If,EqualTo,Count,CollectList,CollectSet,AggregateExpression)]
[2021-08-04T06:38:01.433Z] FAILED ../../src/main/python/hash_aggregate_test.py::test_hash_groupby_collect_with_multi_distinct_fallback[[('a', RepeatSeq(Long)), ('b', RepeatSeq(Decimal(8,3))), ('c', LongRange(not_null))]][IGNORE_ORDER({'local': True}), INCOMPAT, APPROXIMATE_FLOAT, ALLOW_NON_GPU(SortAggregateExec,SortArray,Alias,Literal,First,If,EqualTo,Count,CollectList,CollectSet,AggregateExpression)]
[2021-08-04T06:38:01.433Z] FAILED ../../src/main/python/hash_aggregate_test.py::test_hash_groupby_collect_with_multi_distinct_fallback[[('a', RepeatSeq(Long)), ('b', RepeatSeq(Float)), ('c', LongRange(not_null))]1][IGNORE_ORDER({'local': True}), INCOMPAT, APPROXIMATE_FLOAT, ALLOW_NON_GPU(SortAggregateExec,SortArray,Alias,Literal,First,If,EqualTo,Count,CollectList,CollectSet,AggregateExpression)]
```

detailed log,
```
[2021-08-04T06:38:01.430Z] ----------------------------- Captured stdout call -----------------------------
[2021-08-04T06:38:01.430Z] ### CPU RUN ###
[2021-08-04T06:38:01.430Z] ### GPU RUN ###
[2021-08-04T06:38:01.430Z] [31m[1m_ test_hash_groupby_collect_with_multi_distinct_fallback[[('a', RepeatSeq(Long)), ('b', RepeatSeq(Float)), ('c', LongRange(not_null))]1] _[0m
[2021-08-04T06:38:01.430Z] [gw3] linux -- Python 3.8.8 /databricks/conda/envs/databricks-ml-gpu/bin/python
[2021-08-04T06:38:01.430Z] 
[2021-08-04T06:38:01.430Z] data_gen = [('a', RepeatSeq(Long)), ('b', RepeatSeq(Float)), ('c', LongRange(not_null))]
[2021-08-04T06:38:01.430Z] 
[2021-08-04T06:38:01.430Z]     [37m@approximate_float[39;49;00m
[2021-08-04T06:38:01.430Z]     [37m@ignore_order[39;49;00m(local=[94mTrue[39;49;00m)
[2021-08-04T06:38:01.430Z]     [37m@allow_non_gpu[39;49;00m([33m'[39;49;00m[33mSortAggregateExec[39;49;00m[33m'[39;49;00m,
[2021-08-04T06:38:01.430Z]                    [33m'[39;49;00m[33mSortArray[39;49;00m[33m'[39;49;00m, [33m'[39;49;00m[33mAlias[39;49;00m[33m'[39;49;00m, [33m'[39;49;00m[33mLiteral[39;49;00m[33m'[39;49;00m, [33m'[39;49;00m[33mFirst[39;49;00m[33m'[39;49;00m, [33m'[39;49;00m[33mIf[39;49;00m[33m'[39;49;00m, [33m'[39;49;00m[33mEqualTo[39;49;00m[33m'[39;49;00m, [33m'[39;49;00m[33mCount[39;49;00m[33m'[39;49;00m,
[2021-08-04T06:38:01.431Z]                    [33m'[39;49;00m[33mCollectList[39;49;00m[33m'[39;49;00m, [33m'[39;49;00m[33mCollectSet[39;49;00m[33m'[39;49;00m, [33m'[39;49;00m[33mAggregateExpression[39;49;00m[33m'[39;49;00m)
[2021-08-04T06:38:01.431Z]     [37m@incompat[39;49;00m
[2021-08-04T06:38:01.431Z]     [37m@pytest[39;49;00m.mark.parametrize([33m'[39;49;00m[33mdata_gen[39;49;00m[33m'[39;49;00m, _gen_data_for_collect_op, ids=idfn)
[2021-08-04T06:38:01.431Z]     [94mdef[39;49;00m [92mtest_hash_groupby_collect_with_multi_distinct_fallback[39;49;00m(data_gen):
[2021-08-04T06:38:01.431Z]         [94mdef[39;49;00m [92mspark_fn[39;49;00m(spark_session):
[2021-08-04T06:38:01.431Z]             [94mreturn[39;49;00m gen_df(spark_session, data_gen, length=[94m100[39;49;00m).groupby([33m'[39;49;00m[33ma[39;49;00m[33m'[39;49;00m).agg(
[2021-08-04T06:38:01.431Z]                 f.sort_array(f.collect_list([33m'[39;49;00m[33mb[39;49;00m[33m'[39;49;00m)),
[2021-08-04T06:38:01.431Z]                 f.sort_array(f.collect_set([33m'[39;49;00m[33mb[39;49;00m[33m'[39;49;00m)),
[2021-08-04T06:38:01.431Z]                 f.countDistinct([33m'[39;49;00m[33mb[39;49;00m[33m'[39;49;00m),
[2021-08-04T06:38:01.431Z]                 f.countDistinct([33m'[39;49;00m[33mc[39;49;00m[33m'[39;49;00m))
[2021-08-04T06:38:01.431Z] >       assert_gpu_fallback_collect(func=spark_fn, cpu_fallback_class_name=[33m'[39;49;00m[33mSortAggregateExec[39;49;00m[33m'[39;49;00m)
[2021-08-04T06:38:01.431Z] 
[2021-08-04T06:38:01.431Z] [1m[31m../../src/main/python/hash_aggregate_test.py[0m:476: 
[2021-08-04T06:38:01.431Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[2021-08-04T06:38:01.431Z] [1m[31m../../src/main/python/asserts.py[0m:364: in assert_gpu_fallback_collect
[2021-08-04T06:38:01.431Z]     from_gpu, gpu_df = with_gpu_session(bring_back, conf=conf)
[2021-08-04T06:38:01.431Z] [1m[31m../../src/main/python/spark_session.py[0m:105: in with_gpu_session
[2021-08-04T06:38:01.431Z]     [94mreturn[39;49;00m with_spark_session(func, conf=copy)
[2021-08-04T06:38:01.431Z] [1m[31m../../src/main/python/spark_session.py[0m:70: in with_spark_session
[2021-08-04T06:38:01.431Z]     ret = func(_spark)
[2021-08-04T06:38:01.431Z] [1m[31m../../src/main/python/asserts.py[0m:198: in bring_back
[2021-08-04T06:38:01.431Z]     [94mreturn[39;49;00m (df.collect(), df)
[2021-08-04T06:38:01.431Z] [1m[31m/databricks/spark/python/pyspark/sql/dataframe.py[0m:697: in collect
[2021-08-04T06:38:01.431Z]     sock_info = [96mself[39;49;00m._jdf.collectToPython()
[2021-08-04T06:38:01.431Z] [1m[31m/databricks/spark/python/lib/py4j-0.10.9-src.zip/py4j/java_gateway.py[0m:1304: in __call__
[2021-08-04T06:38:01.431Z]     return_value = get_return_value(
[2021-08-04T06:38:01.431Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[2021-08-04T06:38:01.431Z] 
[2021-08-04T06:38:01.431Z] a = ('xro179879', <py4j.java_gateway.GatewayClient object at 0x7fb5a5b8b610>, 'o179876', 'collectToPython')
[2021-08-04T06:38:01.431Z] kw = {}
[2021-08-04T06:38:01.431Z] converted = IllegalArgumentException('The expression coalesce(first(if ((gid#1739491 = 0)) collect_list(`b`)#1739495 else null) ig...:79)\n\tat py4j.GatewayConnection.run(GatewayConnection.java:251)\n\tat java.lang.Thread.run(Thread.java:748)\n', None)
[2021-08-04T06:38:01.431Z] 
[2021-08-04T06:38:01.431Z]     [94mdef[39;49;00m [92mdeco[39;49;00m(*a, **kw):
[2021-08-04T06:38:01.431Z]         [94mtry[39;49;00m:
[2021-08-04T06:38:01.431Z]             [94mreturn[39;49;00m f(*a, **kw)
[2021-08-04T06:38:01.431Z]         [94mexcept[39;49;00m py4j.protocol.Py4JJavaError [94mas[39;49;00m e:
[2021-08-04T06:38:01.431Z]             converted = convert_exception(e.java_exception)
[2021-08-04T06:38:01.431Z]             [94mif[39;49;00m [95mnot[39;49;00m [96misinstance[39;49;00m(converted, UnknownException):
[2021-08-04T06:38:01.431Z]                 [90m# Hide where the exception came from that shows a non-Pythonic[39;49;00m
[2021-08-04T06:38:01.431Z]                 [90m# JVM exception message.[39;49;00m
[2021-08-04T06:38:01.431Z] >               [94mraise[39;49;00m converted [94mfrom[39;49;00m [96mNone[39;49;00m
[2021-08-04T06:38:01.431Z] [1m[31mE               pyspark.sql.utils.IllegalArgumentException: The expression coalesce(first(if ((gid#1739491 = 0)) collect_list(`b`)#1739495 else null) ignore nulls#1739496, []) is not columnar class org.apache.spark.sql.catalyst.expressions.Coalesce[0m
[2021-08-04T06:38:01.431Z] 
[2021-08-04T06:38:01.431Z] [1m[31m/databricks/spark/python/pyspark/sql/utils.py[0m:116: IllegalArgumentException
```

--- a/integration_tests/src/main/python/hash_aggregate_test.py
+++ b/integration_tests/src/main/python/hash_aggregate_test.py
@@ -452,7 +452,16 @@ def test_hash_groupby_collect_with_single_distinct(data_gen):
 @incompat
 @pytest.mark.parametrize('data_gen', _gen_data_for_collect_op, ids=idfn)
 def test_hash_groupby_single_distinct_collect(data_gen):
-    # test distinct collect with other aggregations
+    # test distinct collect
+    sql = """select a,
+                    sort_array(collect_list(distinct b)),
+                    sort_array(collect_set(distinct b))
+            from tbl group by a"""
+    assert_gpu_and_cpu_are_equal_sql(
+        df_fun=lambda spark: gen_df(spark, data_gen, length=100),
+        table_name="tbl", sql=sql)
+
+    # test distinct collect with nonDistinct aggregations
     sql = """select a,
                     sort_array(collect_list(distinct b)),
                     sort_array(collect_set(b)),
@@ -471,7 +480,7 @@ def test_hash_groupby_single_distinct_collect(data_gen):
 @approximate_float
 @ignore_order(local=True)
 @allow_non_gpu('SortAggregateExec',
-               'SortArray', 'Alias', 'Literal', 'First', 'If', 'EqualTo', 'Count',
+               'SortArray', 'Alias', 'Literal', 'First', 'If', 'EqualTo', 'Count', 'Coalesce',
                'CollectList', 'CollectSet', 'AggregateExpression')
 @incompat
 @pytest.mark.parametrize('data_gen', _gen_data_for_collect_op, ids=idfn)

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/aggregate.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/aggregate.scala
@@ -690,7 +690,7 @@ class GpuHashAggregateIterator(
       val inputAttributes = groupingAttributes ++ distinctAttributes ++ nonDistinctAttributes
       GpuBindReferences.bindGpuReferences(inputProjections, inputAttributes)
     } else if (modeInfo.hasFinalMode ||
-      (modeInfo.hasPartialMergeMode && modeInfo.uniqueModes.length == 1)) {
+        (modeInfo.hasPartialMergeMode && modeInfo.uniqueModes.length == 1)) {
       // This block takes care of two possible conditions:
       // 1. The Final stage, including the 2nd stage of NoDistinctAgg and 4th stage of
       // AggWithOneDistinct, which needs no input projections. Because the child outputs are
@@ -699,7 +699,7 @@ class GpuHashAggregateIterator(
       // taking the child outputs as inputs without any projections.
       GpuBindReferences.bindGpuReferences(childAttr.attrs.asInstanceOf[Seq[Expression]], childAttr)
     } else if (modeInfo.hasPartialMode || modeInfo.hasCompleteMode ||
-      modeInfo.uniqueModes.isEmpty) {
+        modeInfo.uniqueModes.isEmpty) {
       // The first aggregation stage which contains AggExpressions (in either Partial or Complete
       // mode). In this case, the input projections are essential.
       // To be specific, there are four conditions matching this case:
@@ -710,7 +710,7 @@ class GpuHashAggregateIterator(
       //
       // In addition, this block also fits for aggregation stages without any AggExpressions.
       val inputProjections: Seq[Expression] = groupingExpressions ++ aggregateExpressions
-        .flatMap(_.aggregateFunction.inputProjection)
+          .flatMap(_.aggregateFunction.inputProjection)
       GpuBindReferences.bindGpuReferences(inputProjections, childAttr)
     } else {
       // This branch should NOT be reached.
@@ -1238,7 +1238,7 @@ case class GpuHashAggregateExec(
   override lazy val additionalMetrics: Map[String, GpuMetric] = Map(
     NUM_TASKS_FALL_BACKED -> createMetric(MODERATE_LEVEL, DESCRIPTION_NUM_TASKS_FALL_BACKED),
     AGG_TIME -> createNanoTimingMetric(MODERATE_LEVEL, DESCRIPTION_AGG_TIME),
-    CONCAT_TIME-> createNanoTimingMetric(MODERATE_LEVEL, DESCRIPTION_CONCAT_TIME),
+    CONCAT_TIME -> createNanoTimingMetric(MODERATE_LEVEL, DESCRIPTION_CONCAT_TIME),
     SORT_TIME -> createNanoTimingMetric(MODERATE_LEVEL, DESCRIPTION_SORT_TIME)
   ) ++ spillMetrics
 
@@ -1327,8 +1327,8 @@ case class GpuHashAggregateExec(
   // Used in de-duping and optimizer rules
   override def producedAttributes: AttributeSet =
     AttributeSet(aggregateAttributes) ++
-      AttributeSet(resultExpressions.diff(groupingExpressions).map(_.toAttribute)) ++
-      AttributeSet(aggregateBufferAttributes)
+        AttributeSet(resultExpressions.diff(groupingExpressions).map(_.toAttribute)) ++
+        AttributeSet(aggregateBufferAttributes)
 
   // AllTuples = distribution with a single partition and all tuples of the dataset are co-located.
   // Clustered = dataset with tuples co-located in the same partition if they share a specific value
@@ -1369,7 +1369,7 @@ case class GpuHashAggregateExec(
       s"GpuHashAggregate(keys=$keyString, functions=$functionString, output=$outputString)"
     } else {
       s"GpuHashAggregate(keys=$keyString, functions=$functionString)," +
-        s" filters=${aggregateExpressions.map(_.filter)})"
+          s" filters=${aggregateExpressions.map(_.filter)})"
     }
   }
   //

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/aggregate.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/aggregate.scala
@@ -79,10 +79,9 @@ object AggregateUtils {
   /**
    * Computes a target input batch size based on the assumption that computation can consume up to
    * 4X the configured batch size.
-   *
-   * @param confTargetSize  user-configured maximum desired batch size
-   * @param inputTypes      input batch schema
-   * @param outputTypes     output batch schema
+   * @param confTargetSize user-configured maximum desired batch size
+   * @param inputTypes input batch schema
+   * @param outputTypes output batch schema
    * @param isReductionOnly true if this is a reduction-only aggregation without grouping
    * @return maximum target batch size to keep computation under the 4X configured batch limit
    */
@@ -93,7 +92,6 @@ object AggregateUtils {
       isReductionOnly: Boolean): Long = {
     def typesToSize(types: Seq[DataType]): Long =
       types.map(GpuBatchUtils.estimateGpuMemory(_, nullable = false, rowCount = 1)).sum
-
     val inputRowSize = typesToSize(inputTypes)
     val outputRowSize = typesToSize(outputTypes)
     // The cudf hash table implementation allocates four 32-bit integers per input row.
@@ -122,8 +120,7 @@ object AggregateUtils {
 
   /**
    * Compute the aggregation modes and aggregate expressions for all aggregation expressions
-   *
-   * @param aggExpressions      the aggregate expressions
+   * @param aggExpressions the aggregate expressions
    * @param aggBufferAttributes attributes to be bound to the aggregate expressions
    */
   def computeAggModeCudfAggregates(
@@ -199,14 +196,14 @@ object AggregateModeInfo {
  * `buildSortFallbackIterator` is used to sort the aggregated batches by the grouping keys and
  * performs a final merge aggregation pass on the sorted batches.
  *
- * @param cbIter                    iterator providing the nput columnar batches
- * @param groupingExpressions       expressions used for producing the grouping keys
- * @param aggregateExpressions      GPU aggregate expressions used to produce the aggregations
- * @param aggregateAttributes       attribute references to each aggregate expression
- * @param resultExpressions         output expression for the aggregation
- * @param childOutput               input attributes to identify the input columns from the input batches
- * @param modeInfo                  identifies which aggregation modes are being used
- * @param metrics                   metrics that will be updated during aggregation
+ * @param cbIter iterator providing the nput columnar batches
+ * @param groupingExpressions expressions used for producing the grouping keys
+ * @param aggregateExpressions GPU aggregate expressions used to produce the aggregations
+ * @param aggregateAttributes attribute references to each aggregate expression
+ * @param resultExpressions output expression for the aggregation
+ * @param childOutput input attributes to identify the input columns from the input batches
+ * @param modeInfo identifies which aggregation modes are being used
+ * @param metrics metrics that will be updated during aggregation
  * @param configuredTargetBatchSize user-specified value for the targeted input batch size
  */
 class GpuHashAggregateIterator(
@@ -308,7 +305,7 @@ class GpuHashAggregateIterator(
   private def computeTargetMergeBatchSize(confTargetSize: Long): Long = {
     val aggregates = boundExpressions.aggModeCudfAggregates.flatMap(_._2)
     val mergedTypes = groupingExpressions.map(_.dataType) ++ aggregates.map(_.dataType)
-    AggregateUtils.computeTargetBatchSize(confTargetSize, mergedTypes, mergedTypes, isReductionOnly)
+    AggregateUtils.computeTargetBatchSize(confTargetSize, mergedTypes, mergedTypes,isReductionOnly)
   }
 
   /** Aggregate all input batches and place the results in the aggregatedBatches queue. */
@@ -368,7 +365,6 @@ class GpuHashAggregateIterator(
 
   /**
    * Perform a single pass over the aggregated batches attempting to merge adjacent batches.
-   *
    * @return true if at least one merge operation occurred
    */
   private def mergePass(): Boolean = {
@@ -422,7 +418,6 @@ class GpuHashAggregateIterator(
   /**
    * Concatenate batches together and perform a merge aggregation on the result. The input batches
    * will be closed as part of this operation.
-   *
    * @param batches batches to concatenate and merge aggregate
    * @return lazy spillable batch which has NOT been marked spillable
    */
@@ -531,8 +526,8 @@ class GpuHashAggregateIterator(
   /**
    * Project a merged aggregated batch result to the layout that Spark expects
    * i.e.: select avg(foo) from bar group by baz will produce:
-   * Partial mode: 3 columns => [bar, sum(foo) as sum_foo, count(foo) as count_foo]
-   * Final mode:   2 columns => [bar, sum(sum_foo) / sum(count_foo)]
+   *  Partial mode: 3 columns => [bar, sum(foo) as sum_foo, count(foo) as count_foo]
+   *  Final mode:   2 columns => [bar, sum(sum_foo) / sum(count_foo)]
    */
   private def finalProjectBatch(batch: ColumnarBatch): ColumnarBatch = {
     val aggTime = metrics.computeAggTime
@@ -589,7 +584,6 @@ class GpuHashAggregateIterator(
 
   /**
    * Concatenates batches by concatenating the corresponding column vectors within the batches.
-   *
    * @note the input batches are not closed as part of this operation
    * @param batchesToConcat batches to concatenate
    * @return concatenated vectors that together represent the concatenated batch result
@@ -696,7 +690,7 @@ class GpuHashAggregateIterator(
       val inputAttributes = groupingAttributes ++ distinctAttributes ++ nonDistinctAttributes
       GpuBindReferences.bindGpuReferences(inputProjections, inputAttributes)
     } else if (modeInfo.hasFinalMode ||
-        (modeInfo.hasPartialMergeMode && modeInfo.uniqueModes.length == 1)) {
+      (modeInfo.hasPartialMergeMode && modeInfo.uniqueModes.length == 1)) {
       // This block takes care of two possible conditions:
       // 1. The Final stage, including the 2nd stage of NoDistinctAgg and 4th stage of
       // AggWithOneDistinct, which needs no input projections. Because the child outputs are
@@ -705,7 +699,7 @@ class GpuHashAggregateIterator(
       // taking the child outputs as inputs without any projections.
       GpuBindReferences.bindGpuReferences(childAttr.attrs.asInstanceOf[Seq[Expression]], childAttr)
     } else if (modeInfo.hasPartialMode || modeInfo.hasCompleteMode ||
-        modeInfo.uniqueModes.isEmpty) {
+      modeInfo.uniqueModes.isEmpty) {
       // The first aggregation stage which contains AggExpressions (in either Partial or Complete
       // mode). In this case, the input projections are essential.
       // To be specific, there are four conditions matching this case:
@@ -716,7 +710,7 @@ class GpuHashAggregateIterator(
       //
       // In addition, this block also fits for aggregation stages without any AggExpressions.
       val inputProjections: Seq[Expression] = groupingExpressions ++ aggregateExpressions
-          .flatMap(_.aggregateFunction.inputProjection)
+        .flatMap(_.aggregateFunction.inputProjection)
       GpuBindReferences.bindGpuReferences(inputProjections, childAttr)
     } else {
       // This branch should NOT be reached.
@@ -761,16 +755,15 @@ class GpuHashAggregateIterator(
 
   /**
    * Compute the aggregations on the projected input columns.
-   *
    * @param toAggregateCvs column vectors representing the input batch to aggregate
-   * @param merge          true indicates a merge aggregation should be performed
-   * @param isSorted       true indicates the data is already sorted by the grouping keys
+   * @param merge true indicates a merge aggregation should be performed
+   * @param isSorted true indicates the data is already sorted by the grouping keys
    * @return aggregated batch
    */
   private def computeAggregate(
       toAggregateCvs: Seq[GpuColumnVector],
       merge: Boolean,
-      isSorted: Boolean = false): ColumnarBatch = {
+      isSorted: Boolean = false): ColumnarBatch  = {
     val aggModeCudfAggregates = boundExpressions.aggModeCudfAggregates
     val computeAggTime = metrics.computeAggTime
     withResource(new NvtxWithMetrics("computeAggregate", NvtxColor.CYAN, computeAggTime)) { _ =>
@@ -900,12 +893,12 @@ abstract class GpuBaseAggregateMeta[INPUT <: SparkPlan](
 
   override def tagPlanForGpu(): Unit = {
     agg.groupingExpressions
-        .find(_.dataType match {
-          case _@(ArrayType(_, _) | MapType(_, _, _)) | _@StructType(_) => true
-          case _ => false
-        })
-        .foreach(_ =>
-          willNotWorkOnGpu("Nested types in grouping expressions are not supported"))
+      .find(_.dataType match {
+        case _@(ArrayType(_, _) | MapType(_, _, _)) | _@StructType(_) => true
+        case _ => false
+      })
+      .foreach(_ =>
+        willNotWorkOnGpu("Nested types in grouping expressions are not supported"))
     if (agg.resultExpressions.isEmpty) {
       willNotWorkOnGpu("result expressions is empty")
     }
@@ -1221,14 +1214,14 @@ class GpuObjectHashAggregateExecMeta(
  *
  * @param requiredChildDistributionExpressions this is unchanged by the GPU. It is used in
  *                                             EnsureRequirements to be able to add shuffle nodes
- * @param groupingExpressions                  The expressions that, when applied to the input batch, return the
- *                                             grouping key
- * @param aggregateExpressions                 The GpuAggregateExpression instances for this node
- * @param aggregateAttributes                  References to each GpuAggregateExpression (attribute references)
- * @param resultExpressions                    the expected output expression of this hash aggregate (which this
- *                                             node should project)
- * @param child                                incoming plan (where we get input columns from)
- * @param configuredTargetBatchSize            user-configured maximum device memory size of a batch
+ * @param groupingExpressions The expressions that, when applied to the input batch, return the
+ *                            grouping key
+ * @param aggregateExpressions The GpuAggregateExpression instances for this node
+ * @param aggregateAttributes References to each GpuAggregateExpression (attribute references)
+ * @param resultExpressions the expected output expression of this hash aggregate (which this
+ *                          node should project)
+ * @param child incoming plan (where we get input columns from)
+ * @param configuredTargetBatchSize user-configured maximum device memory size of a batch
  */
 case class GpuHashAggregateExec(
     requiredChildDistributionExpressions: Option[Seq[Expression]],
@@ -1245,7 +1238,7 @@ case class GpuHashAggregateExec(
   override lazy val additionalMetrics: Map[String, GpuMetric] = Map(
     NUM_TASKS_FALL_BACKED -> createMetric(MODERATE_LEVEL, DESCRIPTION_NUM_TASKS_FALL_BACKED),
     AGG_TIME -> createNanoTimingMetric(MODERATE_LEVEL, DESCRIPTION_AGG_TIME),
-    CONCAT_TIME -> createNanoTimingMetric(MODERATE_LEVEL, DESCRIPTION_CONCAT_TIME),
+    CONCAT_TIME-> createNanoTimingMetric(MODERATE_LEVEL, DESCRIPTION_CONCAT_TIME),
     SORT_TIME -> createNanoTimingMetric(MODERATE_LEVEL, DESCRIPTION_SORT_TIME)
   ) ++ spillMetrics
 
@@ -1326,7 +1319,7 @@ case class GpuHashAggregateExec(
 
   protected def replaceAlias(attr: AttributeReference): Option[Attribute] = {
     outputExpressions.collectFirst {
-      case a@Alias(child: AttributeReference, _) if child.semanticEquals(attr) =>
+      case a @ Alias(child: AttributeReference, _) if child.semanticEquals(attr) =>
         a.toAttribute
     }
   }
@@ -1334,8 +1327,8 @@ case class GpuHashAggregateExec(
   // Used in de-duping and optimizer rules
   override def producedAttributes: AttributeSet =
     AttributeSet(aggregateAttributes) ++
-        AttributeSet(resultExpressions.diff(groupingExpressions).map(_.toAttribute)) ++
-        AttributeSet(aggregateBufferAttributes)
+      AttributeSet(resultExpressions.diff(groupingExpressions).map(_.toAttribute)) ++
+      AttributeSet(aggregateBufferAttributes)
 
   // AllTuples = distribution with a single partition and all tuples of the dataset are co-located.
   // Clustered = dataset with tuples co-located in the same partition if they share a specific value
@@ -1358,7 +1351,7 @@ case class GpuHashAggregateExec(
    */
   override lazy val allAttributes: AttributeSeq =
     child.output ++ aggregateBufferAttributes ++ aggregateAttributes ++
-        aggregateExpressions.flatMap(_.aggregateFunction.inputAggBufferAttributes)
+      aggregateExpressions.flatMap(_.aggregateFunction.inputAggBufferAttributes)
 
   override def verboseString(maxFields: Int): String = toString(verbose = true, maxFields)
 
@@ -1376,7 +1369,7 @@ case class GpuHashAggregateExec(
       s"GpuHashAggregate(keys=$keyString, functions=$functionString, output=$outputString)"
     } else {
       s"GpuHashAggregate(keys=$keyString, functions=$functionString)," +
-          s" filters=${aggregateExpressions.map(_.filter)})"
+        s" filters=${aggregateExpressions.map(_.filter)})"
     }
   }
   //

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/aggregate.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/aggregate.scala
@@ -628,21 +628,43 @@ class GpuHashAggregateIterator(
     // boundInputReferences is used to pick out of the input batch the appropriate columns
     // for aggregation.
     //
-    // - PartialMerge with Partial mode: we use the inputProjections
-    //   for Partial and non distinct merge expressions for PartialMerge.
-    // - Final or PartialMerge-only mode: we pick the columns in the order as handed to us.
-    // - Partial or Complete mode: we use the inputProjections
+    // - DistinctAggExpressions with nonDistinctAggExpressions in other mode: we switch the
+    //   position of distinctAttributes and nonDistinctAttributes in childAttr. And we use the
+    //   inputProjections for nonDistinctAggExpressions.
+    // - Final mode, PartialMerge-only mode or no AggExpressions: we pick the columns in the order
+    //   as handed to us.
+    // - Partial mode or Complete mode: we use the inputProjections.
     val boundInputReferences =
     if (modeInfo.uniqueModes.length > 1 && aggregateExpressions.exists(_.isDistinct)) {
-      // The 3rd stage of AggWithOneDistinct, which consists of nonDistinctAggExpressions for merge
-      // and distinctAggExpressions for update. For this stage, we need to switch the position of
-      // distinctAttributes and nonDistinctAttributes.
+      // This block takes care of AggregateExec which contains nonDistinctAggExpressions and
+      // distinctAggExpressions with different AggregateModes. All nonDistinctAggExpressions share
+      // one mode and all distinctAggExpressions are in another mode. The specific mode varies in
+      // different Spark runtimes, so this block applies a general condition to adapt different
+      // runtimes:
       //
-      // The schema of the 2nd stage's outputs:
-      // groupingAttributes ++ distinctAttributes ++ nonDistinctAggBufferAttributes
+      // 1. Apache Spark: The 3rd stage of AggWithOneDistinct
+      // The 3rd stage of AggWithOneDistinct, which consists of for nonDistinctAggExpressions in
+      // PartialMerge mode and distinctAggExpressions in Partial mode. For this stage, we need to
+      // switch the position of distinctAttributes and nonDistinctAttributes if there exists at
+      // least one nonDistinctAggExpression. Because the positions of distinctAttributes are ahead
+      // of nonDistinctAttributes in the output of previous stage, since distinctAttributes are
+      // included in groupExpressions.
+      // To be specific, the schema of the 2nd stage's outputs is:
+      // (groupingAttributes ++ distinctAttributes) ++ nonDistinctAggBufferAttributes
+      // The schema of the 3rd stage's expressions is:
+      // groupingAttributes ++ nonDistinctAggExpressions(PartialMerge) ++
+      // distinctAggExpressions(Partial)
       //
-      // The schema of the 3rd stage's expressions:
-      // nonDistinctMergeAggExpressions ++ distinctPartialAggExpressions
+      // 2. Databricks runtime: The final stage of AggWithOneDistinct
+      // Databricks runtime squeezes the 4-stage AggWithOneDistinct into 2 stages. Basically, it
+      // combines the 1st and 2nd stage into a "Partial" stage; and it combines the 3nd and 4th
+      // stage into a "Merge" stage. Similarly, nonDistinctAggExpressions are ahead of distinct
+      // ones in the layout of "Merge" stage's expressions:
+      // groupingAttributes ++ nonDistinctAggExpressions(Final) ++ DistinctAggExpressions(Complete)
+      // Meanwhile, as Apache Spark, distinctAttributes are ahead of nonDistinctAggBufferAttributes
+      // in the output schema of the "Partial" stage.
+      // Therefore, this block also works on the final stage of AggWithOneDistinct under Databricks
+      // runtime.
 
       val (distinctAggExpressions, nonDistinctAggExpressions) = aggregateExpressions.partition(
         _.isDistinct)
@@ -656,10 +678,11 @@ class GpuHashAggregateIterator(
       val distinctAttributes = childAttr.attrs.slice(
         groupingAttributes.length, childAttr.attrs.length - sizeOfNonDistAttr)
 
-      // With PartialMerge modes, we just pass through corresponding attributes of child plan into
-      // nonDistinctExpressions.
+      // For nonDistinctExpressions, they are in either PartialMerge or Final modes. With either
+      // mode, we just need to pass through childAttr.
       val nonDistinctExpressions = nonDistinctAttributes.asInstanceOf[Seq[Expression]]
-      // With Partial modes, the input projections are necessary for distinctExpressions.
+      // For nonDistinctExpressions, they are in either Final or Complete modes. With either mode,
+      // we need to apply the input projections on these AggExpressions.
       val distinctExpressions = distinctAggExpressions.flatMap(_.aggregateFunction.inputProjection)
 
       // Align the expressions of input projections and input attributes
@@ -669,17 +692,24 @@ class GpuHashAggregateIterator(
     } else if (modeInfo.hasFinalMode ||
       (modeInfo.hasPartialMergeMode && modeInfo.uniqueModes.length == 1) ||
       modeInfo.uniqueModes.isEmpty) {
-      // two possible conditions:
+      // This block takes care of three possible conditions:
       // 1. The Final stage, including the 2nd stage of NoDistinctAgg and 4th stage of
       // AggWithOneDistinct, which needs no input projections. Because the child outputs are
       // internal aggregation buffers, which are aligned for the final stage.
       //
       // 2. The 2nd stage (PartialMerge) of AggWithOneDistinct, which works like the final stage
       // taking the child outputs as inputs without any projections.
+      //
+      // 3. Stages without any AggExpressions (only contain groupingExpressions)
       GpuBindReferences.bindGpuReferences(childAttr.attrs.asInstanceOf[Seq[Expression]], childAttr)
     } else if (modeInfo.hasPartialMode || modeInfo.hasCompleteMode) {
-      // The first aggregation stage (including Partial or Complete or no aggExpression),
-      // whose child node is not an AggregateExec. Therefore, input projections are essential.
+      // The first aggregation stage which contains AggExpressions (in either Partial or Complete
+      // mode). In this case, the input projections are essential.
+      // To be specific, there are three conditions matching this case:
+      // 1. The Partial (1st) stage of NoDistinctAgg
+      // 2. The Partial (1st) stage of AggWithOneDistinct
+      // 3. In Databricks runtime, the "Final" (2nd) stage of AggWithOneDistinct which only contains
+      // DistinctAggExpressions (without any nonDistinctAggExpressions)
       val inputProjections: Seq[Expression] = groupingExpressions ++ aggregateExpressions
           .flatMap(_.aggregateFunction.inputProjection)
       GpuBindReferences.bindGpuReferences(inputProjections, childAttr)

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/aggregate.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/aggregate.scala
@@ -633,10 +633,10 @@ class GpuHashAggregateIterator(
     // - Final or PartialMerge-only mode: we pick the columns in the order as handed to us.
     // - Partial or Complete mode: we use the inputProjections
     val boundInputReferences =
-    if (modeInfo.hasPartialMergeMode && modeInfo.hasPartialMode) {
-      // The 3rd stage of AggWithOneDistinct, which combines (partial) reduce-side
-      // nonDistinctAggExpressions and map-side distinctAggExpressions. For this stage, we need to
-      // switch the position of distinctAttributes and nonDistinctAttributes.
+    if (modeInfo.uniqueModes.length > 1 && aggregateExpressions.exists(_.isDistinct)) {
+      // The 3rd stage of AggWithOneDistinct, which consists of nonDistinctAggExpressions for merge
+      // and distinctAggExpressions for update. For this stage, we need to switch the position of
+      // distinctAttributes and nonDistinctAttributes.
       //
       // The schema of the 2nd stage's outputs:
       // groupingAttributes ++ distinctAttributes ++ nonDistinctAggBufferAttributes
@@ -667,7 +667,8 @@ class GpuHashAggregateIterator(
       val inputAttributes = groupingAttributes ++ distinctAttributes ++ nonDistinctAttributes
       GpuBindReferences.bindGpuReferences(inputProjections, inputAttributes)
     } else if (modeInfo.hasFinalMode ||
-        (modeInfo.hasPartialMergeMode && modeInfo.uniqueModes.length == 1)) {
+      (modeInfo.hasPartialMergeMode && modeInfo.uniqueModes.length == 1) ||
+      modeInfo.uniqueModes.isEmpty) {
       // two possible conditions:
       // 1. The Final stage, including the 2nd stage of NoDistinctAgg and 4th stage of
       // AggWithOneDistinct, which needs no input projections. Because the child outputs are
@@ -676,8 +677,7 @@ class GpuHashAggregateIterator(
       // 2. The 2nd stage (PartialMerge) of AggWithOneDistinct, which works like the final stage
       // taking the child outputs as inputs without any projections.
       GpuBindReferences.bindGpuReferences(childAttr.attrs.asInstanceOf[Seq[Expression]], childAttr)
-    } else if (modeInfo.hasPartialMode || modeInfo.hasCompleteMode ||
-        modeInfo.uniqueModes.isEmpty) {
+    } else if (modeInfo.hasPartialMode || modeInfo.hasCompleteMode) {
       // The first aggregation stage (including Partial or Complete or no aggExpression),
       // whose child node is not an AggregateExec. Therefore, input projections are essential.
       val inputProjections: Seq[Expression] = groupingExpressions ++ aggregateExpressions

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/aggregate.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/aggregate.scala
@@ -79,9 +79,10 @@ object AggregateUtils {
   /**
    * Computes a target input batch size based on the assumption that computation can consume up to
    * 4X the configured batch size.
-   * @param confTargetSize user-configured maximum desired batch size
-   * @param inputTypes input batch schema
-   * @param outputTypes output batch schema
+   *
+   * @param confTargetSize  user-configured maximum desired batch size
+   * @param inputTypes      input batch schema
+   * @param outputTypes     output batch schema
    * @param isReductionOnly true if this is a reduction-only aggregation without grouping
    * @return maximum target batch size to keep computation under the 4X configured batch limit
    */
@@ -92,6 +93,7 @@ object AggregateUtils {
       isReductionOnly: Boolean): Long = {
     def typesToSize(types: Seq[DataType]): Long =
       types.map(GpuBatchUtils.estimateGpuMemory(_, nullable = false, rowCount = 1)).sum
+
     val inputRowSize = typesToSize(inputTypes)
     val outputRowSize = typesToSize(outputTypes)
     // The cudf hash table implementation allocates four 32-bit integers per input row.
@@ -120,7 +122,8 @@ object AggregateUtils {
 
   /**
    * Compute the aggregation modes and aggregate expressions for all aggregation expressions
-   * @param aggExpressions the aggregate expressions
+   *
+   * @param aggExpressions      the aggregate expressions
    * @param aggBufferAttributes attributes to be bound to the aggregate expressions
    */
   def computeAggModeCudfAggregates(
@@ -196,14 +199,14 @@ object AggregateModeInfo {
  * `buildSortFallbackIterator` is used to sort the aggregated batches by the grouping keys and
  * performs a final merge aggregation pass on the sorted batches.
  *
- * @param cbIter iterator providing the nput columnar batches
- * @param groupingExpressions expressions used for producing the grouping keys
- * @param aggregateExpressions GPU aggregate expressions used to produce the aggregations
- * @param aggregateAttributes attribute references to each aggregate expression
- * @param resultExpressions output expression for the aggregation
- * @param childOutput input attributes to identify the input columns from the input batches
- * @param modeInfo identifies which aggregation modes are being used
- * @param metrics metrics that will be updated during aggregation
+ * @param cbIter                    iterator providing the nput columnar batches
+ * @param groupingExpressions       expressions used for producing the grouping keys
+ * @param aggregateExpressions      GPU aggregate expressions used to produce the aggregations
+ * @param aggregateAttributes       attribute references to each aggregate expression
+ * @param resultExpressions         output expression for the aggregation
+ * @param childOutput               input attributes to identify the input columns from the input batches
+ * @param modeInfo                  identifies which aggregation modes are being used
+ * @param metrics                   metrics that will be updated during aggregation
  * @param configuredTargetBatchSize user-specified value for the targeted input batch size
  */
 class GpuHashAggregateIterator(
@@ -305,7 +308,7 @@ class GpuHashAggregateIterator(
   private def computeTargetMergeBatchSize(confTargetSize: Long): Long = {
     val aggregates = boundExpressions.aggModeCudfAggregates.flatMap(_._2)
     val mergedTypes = groupingExpressions.map(_.dataType) ++ aggregates.map(_.dataType)
-    AggregateUtils.computeTargetBatchSize(confTargetSize, mergedTypes, mergedTypes,isReductionOnly)
+    AggregateUtils.computeTargetBatchSize(confTargetSize, mergedTypes, mergedTypes, isReductionOnly)
   }
 
   /** Aggregate all input batches and place the results in the aggregatedBatches queue. */
@@ -365,6 +368,7 @@ class GpuHashAggregateIterator(
 
   /**
    * Perform a single pass over the aggregated batches attempting to merge adjacent batches.
+   *
    * @return true if at least one merge operation occurred
    */
   private def mergePass(): Boolean = {
@@ -418,6 +422,7 @@ class GpuHashAggregateIterator(
   /**
    * Concatenate batches together and perform a merge aggregation on the result. The input batches
    * will be closed as part of this operation.
+   *
    * @param batches batches to concatenate and merge aggregate
    * @return lazy spillable batch which has NOT been marked spillable
    */
@@ -526,8 +531,8 @@ class GpuHashAggregateIterator(
   /**
    * Project a merged aggregated batch result to the layout that Spark expects
    * i.e.: select avg(foo) from bar group by baz will produce:
-   *  Partial mode: 3 columns => [bar, sum(foo) as sum_foo, count(foo) as count_foo]
-   *  Final mode:   2 columns => [bar, sum(sum_foo) / sum(count_foo)]
+   * Partial mode: 3 columns => [bar, sum(foo) as sum_foo, count(foo) as count_foo]
+   * Final mode:   2 columns => [bar, sum(sum_foo) / sum(count_foo)]
    */
   private def finalProjectBatch(batch: ColumnarBatch): ColumnarBatch = {
     val aggTime = metrics.computeAggTime
@@ -584,6 +589,7 @@ class GpuHashAggregateIterator(
 
   /**
    * Concatenates batches by concatenating the corresponding column vectors within the batches.
+   *
    * @note the input batches are not closed as part of this operation
    * @param batchesToConcat batches to concatenate
    * @return concatenated vectors that together represent the concatenated batch result
@@ -690,7 +696,7 @@ class GpuHashAggregateIterator(
       val inputAttributes = groupingAttributes ++ distinctAttributes ++ nonDistinctAttributes
       GpuBindReferences.bindGpuReferences(inputProjections, inputAttributes)
     } else if (modeInfo.hasFinalMode ||
-      (modeInfo.hasPartialMergeMode && modeInfo.uniqueModes.length == 1)) {
+        (modeInfo.hasPartialMergeMode && modeInfo.uniqueModes.length == 1)) {
       // This block takes care of two possible conditions:
       // 1. The Final stage, including the 2nd stage of NoDistinctAgg and 4th stage of
       // AggWithOneDistinct, which needs no input projections. Because the child outputs are
@@ -699,7 +705,7 @@ class GpuHashAggregateIterator(
       // taking the child outputs as inputs without any projections.
       GpuBindReferences.bindGpuReferences(childAttr.attrs.asInstanceOf[Seq[Expression]], childAttr)
     } else if (modeInfo.hasPartialMode || modeInfo.hasCompleteMode ||
-      modeInfo.uniqueModes.isEmpty) {
+        modeInfo.uniqueModes.isEmpty) {
       // The first aggregation stage which contains AggExpressions (in either Partial or Complete
       // mode). In this case, the input projections are essential.
       // To be specific, there are four conditions matching this case:
@@ -710,7 +716,7 @@ class GpuHashAggregateIterator(
       //
       // In addition, this block also fits for aggregation stages without any AggExpressions.
       val inputProjections: Seq[Expression] = groupingExpressions ++ aggregateExpressions
-        .flatMap(_.aggregateFunction.inputProjection)
+          .flatMap(_.aggregateFunction.inputProjection)
       GpuBindReferences.bindGpuReferences(inputProjections, childAttr)
     } else {
       // This branch should NOT be reached.
@@ -755,15 +761,16 @@ class GpuHashAggregateIterator(
 
   /**
    * Compute the aggregations on the projected input columns.
+   *
    * @param toAggregateCvs column vectors representing the input batch to aggregate
-   * @param merge true indicates a merge aggregation should be performed
-   * @param isSorted true indicates the data is already sorted by the grouping keys
+   * @param merge          true indicates a merge aggregation should be performed
+   * @param isSorted       true indicates the data is already sorted by the grouping keys
    * @return aggregated batch
    */
   private def computeAggregate(
       toAggregateCvs: Seq[GpuColumnVector],
       merge: Boolean,
-      isSorted: Boolean = false): ColumnarBatch  = {
+      isSorted: Boolean = false): ColumnarBatch = {
     val aggModeCudfAggregates = boundExpressions.aggModeCudfAggregates
     val computeAggTime = metrics.computeAggTime
     withResource(new NvtxWithMetrics("computeAggregate", NvtxColor.CYAN, computeAggTime)) { _ =>
@@ -893,12 +900,12 @@ abstract class GpuBaseAggregateMeta[INPUT <: SparkPlan](
 
   override def tagPlanForGpu(): Unit = {
     agg.groupingExpressions
-      .find(_.dataType match {
-        case _@(ArrayType(_, _) | MapType(_, _, _)) | _@StructType(_) => true
-        case _ => false
-      })
-      .foreach(_ =>
-        willNotWorkOnGpu("Nested types in grouping expressions are not supported"))
+        .find(_.dataType match {
+          case _@(ArrayType(_, _) | MapType(_, _, _)) | _@StructType(_) => true
+          case _ => false
+        })
+        .foreach(_ =>
+          willNotWorkOnGpu("Nested types in grouping expressions are not supported"))
     if (agg.resultExpressions.isEmpty) {
       willNotWorkOnGpu("result expressions is empty")
     }
@@ -1214,14 +1221,14 @@ class GpuObjectHashAggregateExecMeta(
  *
  * @param requiredChildDistributionExpressions this is unchanged by the GPU. It is used in
  *                                             EnsureRequirements to be able to add shuffle nodes
- * @param groupingExpressions The expressions that, when applied to the input batch, return the
- *                            grouping key
- * @param aggregateExpressions The GpuAggregateExpression instances for this node
- * @param aggregateAttributes References to each GpuAggregateExpression (attribute references)
- * @param resultExpressions the expected output expression of this hash aggregate (which this
- *                          node should project)
- * @param child incoming plan (where we get input columns from)
- * @param configuredTargetBatchSize user-configured maximum device memory size of a batch
+ * @param groupingExpressions                  The expressions that, when applied to the input batch, return the
+ *                                             grouping key
+ * @param aggregateExpressions                 The GpuAggregateExpression instances for this node
+ * @param aggregateAttributes                  References to each GpuAggregateExpression (attribute references)
+ * @param resultExpressions                    the expected output expression of this hash aggregate (which this
+ *                                             node should project)
+ * @param child                                incoming plan (where we get input columns from)
+ * @param configuredTargetBatchSize            user-configured maximum device memory size of a batch
  */
 case class GpuHashAggregateExec(
     requiredChildDistributionExpressions: Option[Seq[Expression]],
@@ -1238,7 +1245,7 @@ case class GpuHashAggregateExec(
   override lazy val additionalMetrics: Map[String, GpuMetric] = Map(
     NUM_TASKS_FALL_BACKED -> createMetric(MODERATE_LEVEL, DESCRIPTION_NUM_TASKS_FALL_BACKED),
     AGG_TIME -> createNanoTimingMetric(MODERATE_LEVEL, DESCRIPTION_AGG_TIME),
-    CONCAT_TIME-> createNanoTimingMetric(MODERATE_LEVEL, DESCRIPTION_CONCAT_TIME),
+    CONCAT_TIME -> createNanoTimingMetric(MODERATE_LEVEL, DESCRIPTION_CONCAT_TIME),
     SORT_TIME -> createNanoTimingMetric(MODERATE_LEVEL, DESCRIPTION_SORT_TIME)
   ) ++ spillMetrics
 
@@ -1319,7 +1326,7 @@ case class GpuHashAggregateExec(
 
   protected def replaceAlias(attr: AttributeReference): Option[Attribute] = {
     outputExpressions.collectFirst {
-      case a @ Alias(child: AttributeReference, _) if child.semanticEquals(attr) =>
+      case a@Alias(child: AttributeReference, _) if child.semanticEquals(attr) =>
         a.toAttribute
     }
   }
@@ -1327,8 +1334,8 @@ case class GpuHashAggregateExec(
   // Used in de-duping and optimizer rules
   override def producedAttributes: AttributeSet =
     AttributeSet(aggregateAttributes) ++
-      AttributeSet(resultExpressions.diff(groupingExpressions).map(_.toAttribute)) ++
-      AttributeSet(aggregateBufferAttributes)
+        AttributeSet(resultExpressions.diff(groupingExpressions).map(_.toAttribute)) ++
+        AttributeSet(aggregateBufferAttributes)
 
   // AllTuples = distribution with a single partition and all tuples of the dataset are co-located.
   // Clustered = dataset with tuples co-located in the same partition if they share a specific value
@@ -1351,7 +1358,7 @@ case class GpuHashAggregateExec(
    */
   override lazy val allAttributes: AttributeSeq =
     child.output ++ aggregateBufferAttributes ++ aggregateAttributes ++
-      aggregateExpressions.flatMap(_.aggregateFunction.inputAggBufferAttributes)
+        aggregateExpressions.flatMap(_.aggregateFunction.inputAggBufferAttributes)
 
   override def verboseString(maxFields: Int): String = toString(verbose = true, maxFields)
 
@@ -1369,7 +1376,7 @@ case class GpuHashAggregateExec(
       s"GpuHashAggregate(keys=$keyString, functions=$functionString, output=$outputString)"
     } else {
       s"GpuHashAggregate(keys=$keyString, functions=$functionString)," +
-        s" filters=${aggregateExpressions.map(_.filter)})"
+          s" filters=${aggregateExpressions.map(_.filter)})"
     }
   }
   //

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/aggregate.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/aggregate.scala
@@ -690,28 +690,27 @@ class GpuHashAggregateIterator(
       val inputAttributes = groupingAttributes ++ distinctAttributes ++ nonDistinctAttributes
       GpuBindReferences.bindGpuReferences(inputProjections, inputAttributes)
     } else if (modeInfo.hasFinalMode ||
-      (modeInfo.hasPartialMergeMode && modeInfo.uniqueModes.length == 1) ||
-      modeInfo.uniqueModes.isEmpty) {
-      // This block takes care of three possible conditions:
+      (modeInfo.hasPartialMergeMode && modeInfo.uniqueModes.length == 1)) {
+      // This block takes care of two possible conditions:
       // 1. The Final stage, including the 2nd stage of NoDistinctAgg and 4th stage of
       // AggWithOneDistinct, which needs no input projections. Because the child outputs are
       // internal aggregation buffers, which are aligned for the final stage.
-      //
       // 2. The 2nd stage (PartialMerge) of AggWithOneDistinct, which works like the final stage
       // taking the child outputs as inputs without any projections.
-      //
-      // 3. Stages without any AggExpressions (only contain groupingExpressions)
       GpuBindReferences.bindGpuReferences(childAttr.attrs.asInstanceOf[Seq[Expression]], childAttr)
-    } else if (modeInfo.hasPartialMode || modeInfo.hasCompleteMode) {
+    } else if (modeInfo.hasPartialMode || modeInfo.hasCompleteMode ||
+      modeInfo.uniqueModes.isEmpty) {
       // The first aggregation stage which contains AggExpressions (in either Partial or Complete
       // mode). In this case, the input projections are essential.
-      // To be specific, there are three conditions matching this case:
+      // To be specific, there are four conditions matching this case:
       // 1. The Partial (1st) stage of NoDistinctAgg
       // 2. The Partial (1st) stage of AggWithOneDistinct
       // 3. In Databricks runtime, the "Final" (2nd) stage of AggWithOneDistinct which only contains
       // DistinctAggExpressions (without any nonDistinctAggExpressions)
+      //
+      // In addition, this block also fits for aggregation stages without any AggExpressions.
       val inputProjections: Seq[Expression] = groupingExpressions ++ aggregateExpressions
-          .flatMap(_.aggregateFunction.inputProjection)
+        .flatMap(_.aggregateFunction.inputProjection)
       GpuBindReferences.bindGpuReferences(inputProjections, childAttr)
     } else {
       // This branch should NOT be reached.

